### PR TITLE
Update after Erik's feedback

### PR DESCRIPTION
**Pull request by rchiodo**
Hope this works. Please feel free to comment.

### CODE REVIEW

File: pep-0722.rst [hunk 1]
```diff
@@ -1,6 +1,6 @@
 PEP: 722
 Title: Stricter Type Guards
-Author: Rich Chiodo <rchiodo at microsoft.com>, Eric Traut <eric at traut.com>
+Author: Rich Chiodo <rchiodo at microsoft.com>, Eric Traut <erictr at microsoft.com>, Erik De Bonte <erikd at microsoft.com>
```
Comment (debonte) [inline pep-0722.rst:3]: You should use a non-microsoft email address for Eric.

Comment (debonte) [inline pep-0722.rst:3]: Not sure I deserve to be in the authors list. Your call. You'll get feedback from lots of people and probably don't want to add all feedback providers to this list.

File: pep-0722.rst [hunk 2]
```diff
@@ -17,87 +17,99 @@ Resolution:
 Abstract
 ========
 
-This PEP further refines `TypeGuards <typeguards_>`__ to
-indicate when negative type narrowing is deemed safe.
-
-[I'd suggest mentioning PEP 647 explicitly here rather than having the opaque link. You can link to a PEP in RST using :pep:`647` ]
-[I think more context is needed here for readers to understand what "negative" means. Maybe one sentence explaining what typeguards currently do and then another about the negative issue.]
+:pep:`647` created a special return type annotation ``TypeGuard`` that allowed
+type checkers to narrow types.
```
Comment (debonte) [inline pep-0722.rst]: I think a little more context would still be helpful here.

```suggestion
:pep:`647` introduced the concept of ``TypeGuard`` functions which return true if their input parameter matches their target type. For example, a function that returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's input parameter is a ``str``. This allows type checkers to narrow types in this positive case.
```

File: pep-0722.rst [hunk 3]
```diff
@@ -17,87 +17,99 @@ Resolution:
...
+This PEP further refines :pep:`647` by allowing type checkers to narrow types
+even further when a ``TypeGuard`` function returns false.
 
 Motivation
 ==========
 
-`TypeGuards <typeguards_>`__ are used throughout python
-libraries but cannot be used to determine the negative case:
+`TypeGuards <typeguards_>`__ are used throughout Python libraries to allow a
+type checker to narrow what the type of something is when the ``TypeGuard``
```
Comment (debonte) [inline pep-0722.rst]:
```suggestion
type checker to narrow the type of something when the ``TypeGuard``
```

File: pep-0722.rst [hunk 4]
```diff
@@ -17,87 +17,99 @@ Resolution:
...
 
-[Again, more context is needed for the user to understand what "the negative case" means.]
-[Also what does "determine the negative case" mean? Maybe something like "narrow the type in the negative case" would be more clear? Also see the use of that phrase below the code block.]
-[python should be capitalized]
+    def is_str(val: str | int) -> TypeGuard[str]:
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
```
Comment (debonte) [inline pep-0722.rst]: Maybe this first code block could be removed since the second one is identical except for adding the `else` block?

File: pep-0722.rst [hunk 5]
```diff
@@ -17,87 +17,99 @@ Resolution:
...
     def is_str(val: str | int) -> TypeGuard[str]:
         return isinstance(val, str)
 
     def func(val: str | int):
         if is_str(val):
-            reveal_type(val) # str
+            # Type checkers can assume val is a 'str' in this branch
         else:
-            reveal_type(val) # int
+            # Type checkers can assume val is an 'int' in this branch
```
Comment (debonte) [inline pep-0722.rst:66]: Maybe this code block is also not necessary and you could say something like:

> This allows a type checker to narrow the type of ``val`` to ``int`` in the ``else`` block of the example above.

My main concern is readability. Very little has changed in each of these code blocks, but the user needs to scan the whole thing and compare against the previous code block to make sure they're not missing something important.

File: pep-0722.rst [hunk 6]
```diff
@@ -17,87 +17,99 @@ Resolution:
...
+
+Since the ``TypeGuard`` type (or output type) is a subtype of the input argument
+type, a type checker can determine that the only possible type in the ``else``
+is the other type in the Union. In this example, it is safe to assume that if
```
Comment (debonte) [inline pep-0722.rst]: ``Union`` or lowercase union.

File: pep-0722.rst [hunk 7]
```diff
@@ -17,87 +17,136 @@ Resolution:
...
     def func(val: list[int | str]):
         if is_str_list(val):
-            reveal_type(val) # list[str]
+            # Type checker assumes list[str] here
+        else:
+            # Type checker cannot assume list[int] here
+
+Since ``list`` is invariant, it doesn't have any subtypes. This means type
+checkers cannot narrow the type to ``list[int]`` in the false case.
+``list[str]`` is not a subtype of ``list[str | int]``.
```
Comment (debonte) [inline pep-0722.rst]: Suggest moving the last sentence after the first sentence.

```suggestion
Since ``list`` is invariant, it doesn't have any subtypes. ``list[str]`` is not a subtype of ``list[str | int]``. This means type
checkers cannot narrow the type to ``list[int]`` in the false case.
```

File: pep-0722.rst [hunk 8]
```diff
@@ -17,87 +17,136 @@ Resolution:
...
+
+Type checkers should not assume any narrowing in the false case when the
+``TypeGuard`` type is not a subtype of the input argument type.
```
Comment (debonte) [inline pep-0722.rst]: Maybe "TypeGuard type argument" is a good way to refer to the "output type"?

```suggestion
``TypeGuard`` type argument is not a subtype of the input argument type. 
```

File: pep-0722.rst [hunk 9]
```diff
@@ -17,87 +17,136 @@ Resolution:
...
+.. code-block:: python
+
+    PosInt = NewType('PosInt', int)
+
+    def is_positive_int(val: PosInt | int | str) -> TypeGuard[PosInt]:
+        return isinstance(val, int) and val > 0
+
+    def func(val: int | str):
+        if is_positive_int(val):
+            # Type checker assumes PosInt here
```
Comment (debonte) [inline pep-0722.rst:147]: > Type checker assumes PosInt here

This is true even though `val` is `int | str` (i.e. it doesn't include `PosInt`)?

File: pep-0722.rst [hunk 10]
```diff
@@ -17,87 +17,136 @@ Resolution:
...
-["the other input type(s)" -- There's only one input type. It's a Union. Suggest rephrasing this. I'm not sure if talking about the types using set theory (input -- output) would make this more clear (or more generic) or worse.]
+Since the ``TypeGuard`` type (or output type) is a subtype of the input argument
+type, a type checker can determine that the only possible type in the ``else``
+is the other type in the Union. In this example, it is safe to assume that if
+``is_str`` returns false, then type of the ``val`` argument is an ``int``.
 
-If the output type is not a *strict* subtype of the input type,
-the negative cannot be assumed to be the intuitive opposite:
-["intuitive opposite" -- opposite is the incorrect term here and I think intuition doesn't belong in a PEP :)]
+Unsafe Narrowing
```
Comment (debonte) [inline pep-0722.rst:86]: It seems weird to me that the "Unsafe narrowing" and "Creating invalid narrowing" sections are within "Motivation"

File: pep-0722.rst [hunk 11]
```diff
@@ -108,49 +157,50 @@ Backwards Compatibility
 For preexisting code this should require no changes, but should simplify this
 use case here:
```
Comment (debonte) [inline pep-0722.rst]:
```suggestion
For preexisting code this should require no changes, but will allow
use cases such as the one below to be simplified:
```

File: pep-0722.rst [hunk 12]
```diff
@@ -108,49 +157,50 @@ Backwards Compatibility
...
+        # Do stuff assuming x is a 'B'
         return
 
-["after_is_A" is confusing me -- is there a better name? "is_not_A"?]
-[Can/should you use PEP 695 syntax for the TypeVars?]
 
-becomes this instead
-["becomes this instead" is not a grammatically correct continuation of the sentence before the first code block. Maybe rephrase the sentence to "Preexisting code should require no changes, but code like this...can be simplified to this:"]
-[Add comments in these code blocks showing the expected inferred type as you did above? I think then you won't need the reveal_type calls?]
+This use case becomes this instead:
```
Comment (debonte) [inline pep-0722.rst]: With this proposed change, the code above continues to work but could be simplified by removing the assertion that `x` is of type `B` in the negative case:

File: pep-0722.rst (hunk 1 above)
Comment (rchiodo) [reply pep-0722.rst:3]: Hmm, okay. You are the first person to look at it though :)

Comment (rchiodo) [reply pep-0722.rst:3]: Unfortunately building crashes if that's the case. The pep build system only allows one e-mail per name. So I'd have to update all of the other PEPs that Eric is mentioned in.

File: pep-0722.rst (hunk 10 above)
Comment (rchiodo) [reply pep-0722.rst:86]: Where do you think they should go? Rejected ideas?

File: pep-0722.rst (hunk 9 above)
Comment (rchiodo) [reply pep-0722.rst:147]: Yes that's true. This works because PosInt is an int.

File: pep-0722.rst (hunk 6 above)
Comment (rchiodo) [reply pep-0722.rst]: Oh you mean put it in quotes. Okay.

File: pep-0722.rst (hunk 4 above)
Comment (rchiodo) [reply pep-0722.rst]: I think it should be there cause I didn't want to introduce the else case until after the user had seen the if case.

File: pep-0722.rst (hunk 9 above)
Comment (rchiodo) [reply pep-0722.rst:147]: You should be able to try it yourself in the latest pylance.

Create a pyrightconfig.json with this in it:

```json
{
    "enableExperimentalFeatures" : true
}
```

Then the TypeGuard will behave as I described here.

File: pep-0722.rst (hunk 5 above)
Comment (rchiodo) [reply pep-0722.rst:66]: I was hoping to show the evolution of what the 'else' statement could mean. I'm not sure if I remove any of the examples if that's still clear or not.

Perhaps I can just show this?

```python
if is_str(val)
   # Type checkers can assume val is 'str'
else 
   # This is the only part that would change in the subsequent examples
```

File: pep-0722.rst (hunk 10 above)
Comment (rchiodo) [reply pep-0722.rst:86]: In Eric's PEP he does describe some of the situations where TypeGuards don't work in the motivation section:
https://peps.python.org/pep-0647/

File: pep-0722.rst [hunk 13]
```diff
@@ -17,140 +17,196 @@ Resolution:
...
-[I'd suggest mentioning PEP 647 explicitly here rather than having the opaque link. You can link to a PEP in RST using :pep:`647` ]
-[I think more context is needed here for readers to understand what "negative" means. Maybe one sentence explaining what typeguards currently do and then another about the negative issue.]
+:pep:`647` introduced the concept of ``TypeGuard`` functions which return true
+if their input parameter matches their target type. For example, a function that
+returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's
+input parameter is a ``str``. This allows type checkers to narrow types in this
+positive case.
 
+This PEP further refines :pep:`647` by allowing type checkers to narrow types
+even further when a ``TypeGuard`` function returns false.
```
Comment (debonte) [inline pep-0722.rst]:
```suggestion
This PEP further refines :pep:`647` by allowing type checkers to also narrow types
when a ``TypeGuard`` function returns false.
```

File: pep-0722.rst (hunk 4 above)
Comment (debonte) [reply pep-0722.rst]: Just want to make sure we're talking about the same thing. I was envisioning the following. And maybe to mention in the `if` block comment that the "val is a str" behavior is already defined by PEP 647.

```text
`TypeGuards` are used throughout Python libraries to allow a
type checker to narrow the type of something when the ``TypeGuard``
returns true. However, in the ``else`` clause, :pep:`647` didn't prescribe
what the type might be:
```

```python
    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:
            # Type here is not narrowed. It is still 'str | int' 
```

File: pep-0722.rst [hunk 14]
```diff
@@ -17,140 +17,196 @@ Resolution:
...
+            # Type checkers can assume val is a 'str' in this branch
         else:
-            reveal_type(val) # str | int
+            # Type here is not narrowed. It is still 'str | int' 
+
 
-This inability to determine the negative case makes ``TypeGuard`` not as useful as
-it could be.
+This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype
+of the type of the first input argument, then the false case can be further
```
Comment (debonte) [inline pep-0722.rst]: "first input argument" -- are type guards allowed to have more than one parameter?

Comment (debonte) [inline pep-0722.rst]: IIRC, an argument is a value that you pass to a function, so what you're talking about here is a parameter.

```suggestion
of the type of the first input parameter, then the false case can be further
```

File: pep-0722.rst (hunk 5 above)
Comment (debonte) [reply pep-0722.rst:66]: Ok. Thinking about this more, I think having "before this PEP" + "with this PEP" examples makes sense for comparison. But I still think that the first sample is unnecessary. See my other comment.

File: pep-0722.rst (hunk 10 above)
Comment (debonte) [reply pep-0722.rst:86]: I was thinking "Specification". Counter examples don't strike me as part of the *motivation* for this PEP.

File: pep-0722.rst (hunk 9 above)
Comment (debonte) [reply pep-0722.rst:147]: I believe that that's how Pyright works. I guess I'm wondering:

a) Is this correct?

b) What would be the behavior if `PosInt` was unrelated to any of the types in the `val` union? Consider the following example. Within the `if` block, is `val` considered to be `datetime` because `is_datetime` returned `true`? Or is it `Never` because the intersection of `datetime` and `int | str` is empty? Oh, I guess this might be the multiple inheritance thing again. Maybe the `val` passed into `func` is derived from both `int` and `datetime`?

```python
def is_datetime(val: datetime | int | str) -> TypeGuard[datetime]:
    return isinstance(val, datetime)

def func(val: int | str):
    if is_datetime(val):
        # Is val a datetime here or Never?
```

c) Should the PEP call more attention to these issues?

File: pep-0722.rst (hunk 14 above)
Comment (rchiodo) [reply pep-0722.rst]: They are. The typechecker only looks at the first one though.

**Comment by rchiodo** [comment]
> I was thinking "Specification". Counter examples don't strike me as part of the motivation for this PEP.

Sounds good. I'll move it.

File: pep-0722.rst (hunk 10 above)
Comment (rchiodo) [reply pep-0722.rst:86]: Sounds good.

File: pep-0722.rst (hunk 9 above)
Comment (rchiodo) [reply pep-0722.rst:147]: I believe the TypeGuard pep says that if the TypeGuard returns something nonsensical, that's the user's fault. It will just happily presume that's the actual type if the TypeGuard returns true.

This is exactly what Pyright does at the moment.

Comment (rchiodo) [reply pep-0722.rst:147]: It's mentioned here:
https://peps.python.org/pep-0647/#enforcing-strict-narrowing
